### ISSUE TITLE
dict2dzn / arrays indexed by enums not possible?

### ISSUE THREAD

**Issue by Nitrogenycs** [paolodragone/pymzn#24]
Hey again,

In my mzn I have a type definition like this:

```
enum JOB;
array[JOB] of int: job_durations;
```

Now when I feed my data to `minizinc(data={"job_durations": [1,2,3], ...)` then it will result in a data file like this:

`task_durations = array1d(1..3, [1, 2, 3]);` 

This results in an error in MiniZinc:

`MiniZinc: type error: initialisation value for 'job_durations' has invalid type-inst: expected 'array[JOB] of int', actual 'array[int] of int'`

Looking at the code, especially `_dzn_array_nd` and `_index_set` in `marsh.py` it seems that this use case is not supported.

Is there a way to make this work?

When I do things manually I do it like this:

```
job_durations = [ 1, 2, 3 ];
job_flags = 
[| false, false, false,  
 | false, false, false,
 | false, false, false
 |];
```

for 1- and 2d-arrays respectively. That is, I just omit the dimensions and arrayNd stuff.

**Comment by Nitrogenycs** [comment]
The reverse process dzn2dict also seems to have problems with this.

Output of solver:
```
slots = [{Slot1_20180602, Slot2_20180602}]
```

Output I get from solutions in python:

```
{'slots': [20180602, 20180602]}
```

That is, it just seems to drop everything that doesn't look numeric.

**Comment by paolodragone** [comment]
Yeah, the enums are still not supported by PyMzn unfortunately.
They are a bit of a mess to convert to python so I've never actually got into them.
The standard workaround is of course with to map the enum to an array of integers, which you then convert back in the python code when you get the result.
I was actually planning to add support for them, but I decided to wait for the release of MiniZinc 2.2.0 before adding any new major feature.
I'll keep this issue open as a reminder!

**Comment by Nitrogenycs** [comment]
Thank you for the detailed information. My workaround now is this: 

1. I create the input data file myself with jinja2, still based on enums
2. To get the enums in the result I return the results as "json" instead of "dict". This will return only the integer values and no enums.

**Comment by alvarohivery** [comment]
Hi Paolo, 
Thanks for the tool is been really helpful. Is this still an open issue? experiencing the same problem as Nitrogenics, and wondering if there is any feasible workaround yet.

Cheers!

**Comment by paolodragone** [comment]
Hi alvarohivery, yes, this is still an issue unfortunately, I'm planning to give a round of big improvements like this once the MiniZinc 2.2.0 becomes a bit more stable.
The workaround is simply to use integer variables or arrays to encode enum values. In the python code then you'll have the resulting integers mapped to whatever you want.
Hope this helps

**Comment by alvarohivery** [comment]
Hi Paolo, this is similar to what I ended up doing. Thanks for the
response!.

Saving multidimensional boolean arrays is not supported as well right? I
tried using np.arrays with dtype=bool and it did not work directly, but
with arrays of arrays it ended up working.

Thanks for the good work Paolo!

Cheers

**Comment by shuaiwang88** [comment]
Hi Paolo, 
I was trying to duplicate the minizinc example. 
I got the same error for profit, capacity, and consumption similar to this   "expected `array[Products] of int', actual `array[int] of int'" . 

Do we have any work around here? 
```
Products = {ChocolateCake, BananaCake};

profit = array1d(1..2, [400, 450]);

Resources = {Butter, Banana, Flour, Sugar, Cocoa};

capacity = array1d(1..5, [4000, 6, 2000, 500, 500]);

consumption = array2d(1..2, 1..5, [250.0, 75.0, 0.0, 0.0, 150.0, 2.0, 100.0, 200.0,
    150.0, 75.0]);


```

**Comment by paolodragone** [comment]
Fixed in 0.18.1.
